updater/updater.go

```diff
@@ -0,0 +1,345 @@
+		return fmt.Errorf("reporting progress to omaha server, got status %q", app.Status)
+	}
+
+	return nil
+}
+
+// ReportError takes an optional errorCode and reports
+// that an error occured during the installation process,
+// the optional errorCode can be used to send custom
+// error codes to the server. This error code can then be
+// used to trace out errors custom to the application
+// installation process.
+func (u *updater) ReportError(ctx context.Context, errorCode *int) error {
+	errorEvent := progressToEventRequest(ProgressError)
+	if errorCode != nil {
+		errorEvent.ErrorCode = *errorCode
+	}
+
+	resp, err := u.SendOmahaEvent(ctx, errorEvent)
+	if err != nil {
+		return fmt.Errorf("sending progress error event to omaha server: %w", err)
+	}
+
+	app := resp.GetApp(u.appID)
+	if app.Status != "ok" {
+		return fmt.Errorf("reporting progress error to omaha server, got status %q", app.Status)
+	}
+
+	return nil
+}
+
+// CompleteUpdate sends an ProgressUpdateComplete event to the omaha server
+// and sets the version in the UpdateInfo as the current version of
+// the instance in the updater.
+func (u *updater) CompleteUpdate(ctx context.Context, info *UpdateInfo) error {
+	if info == nil {
+		return errors.New("invalid UpdateInfo")
+	}
+
+	version := info.Version
+	if version == "" {
+		return fmt.Errorf("invalid version, can't report complete event to omaha server")
+	}
+
+	err := u.ReportProgress(ctx, ProgressUpdateComplete)
+	if err != nil {
+		return fmt.Errorf("reporting ProgressUpdateComplete to omaha server: %w", err)
+	}
+
+	u.SetInstanceVersion(version)
+	return nil
+}
+
+// SendOmahaRequest sends the event request to the omaha server
+// and returns the omaha.Response returned by the omaha server.
+func (u *updater) SendOmahaEvent(ctx context.Context, event *omaha.EventRequest) (*omaha.Response, error) {
+	req := u.newAppRequest()
+	app := req.GetApp(u.appID)
+	app.Events = append(app.Events, event)
+
+	return u.SendOmahaRequest(ctx, req)
+}
+
+// InstanceVersion returns the current version of the application.
+func (u *updater) InstanceVersion() string {
+	u.mu.RLock()
+	defer u.mu.RUnlock()
+	return u.instanceVersion
+}
+
+// SetInstanceVersion sets the current instance version
+// of the application to the updater.
+func (u *updater) SetInstanceVersion(version string) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	u.instanceVersion = version
+}
+
+// TryUpdate function takes in an UpdateHandler and performs
+// the complete flow from checking for updates to reporting
+// status etc and returns an error if anything fails in the flow.
+func (u *updater) TryUpdate(ctx context.Context, handler UpdateHandler) error {
+	if handler == nil {
+		return errors.New("invalid UpdateHandler")
+	}
+
+	info, err := u.CheckForUpdates(ctx)
+	if err != nil {
+		return fmt.Errorf("checking for update: %w", err)
+	}
+
+	if !info.HasUpdate {
+		return fmt.Errorf("%w %v, channel %v: %v", NoUpdateError{}, u.appID, u.channel, info.UpdateStatus)
```

Comment:
Maybe we can make those fields a struct field, so they are easily accessible by the client? It seems like a lot of formatting here.
```suggestion
		return NoUpdateError{
		  AppID: u.appID,
		  Channel: u.channel,
		  UpdateStateus: info.UpdateStatus,
		}
```